src/results.jl

```diff
@@ -0,0 +1,130 @@
+"""
+   InternalIntegralResult
+
+Result of computing the internal integral I(phi, beta1, beta2)
+- value: the computed integral value
+- error: error estimate (Nothing if not available)
+"""
+struct InternalIntegralResult{T <: Number}
+    value::T
+    error::T
+end
+
+# Constructor for cases without error estimate
+InternalIntegralResult(value::T) where {T} = InternalIntegralResult(value, T(NaN))
+
+function Base.isapprox(res1::InternalIntegralResult, res2::InternalIntegralResult, kwargs...)
+    return isapprox(res1.value, res2.value, kwargs...)
+end
+
+"""
+    InternalIntegrals
+
+Stores the values of internal integrals.
+
+For definite polarization:
+
+```math
+I_1 = \\int_0^x f(y)\\,\\mathrm{d}y\\\\
+I_2 = \\int_0^x f^2(y)\\,\\mathrm{d}y
+```
+
+For indefinite polarization:
+
+```math
+
+I_11 = \\int_0^x f_x(y)\\,\\mathrm{d}y\\\\
+I_12 = \\int_0^x f_y(y)\\,\\mathrm{d}y\\\\
+I_2 = \\int_0^x f_x^2(y)+ f_y^2(y)\\,\\mathrm{d}y
+```
+"""
+struct InternalIntegrals{T <: Number, N}
+    data::NTuple{N, InternalIntegralResult{T}}
+    function InternalIntegrals(
```

Comment:
This looks like a copy constructor that deduces the inner type `T` of the `InternalIntegrateResult{T}` of data just to construct the same type `NTuple{N, InternalIntegralResult{T}}` in the end again.

Why do you do that?

Reply:
Right, this infers the fundamental element type from the input and exposes it as a type parameter, i.e., `InternalIntegrals{T} where T<:Number`. This allows dispatch on the element type and gives the Julia compiler the freedom to optimize it.